### PR TITLE
Decorate .instance_method to proxy original params

### PR DESCRIPTION
**Pull request by ms-ati**
In Panorama spec helper code 'initialized_double', we observed
that MemoWise interrupts the use of Ruby reflection on method
parameters, by replacing with delegating methods that have overly
generic signatures, like:

    def initialize(...)

    def slow_method(*arg, **kwargs)

After investigation, we had determined that it was not worth
the complexity cost, and potential fragility, to pursue a solution
where we re-create the original method parmeter signatures for our
delegating implementations.

This PR is a minimalistic alternative, which narrowly overrides only
the implementation of {Module#instance_method} so that, in the case
of either #initialize or a method that has had .memo_wise called on it,
we look up the original method and proxy the original result value of
calling UnboundMethod#parameters on it.

### CODE REVIEW

File: spec/proxying_original_method_params_spec.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      it "returns parameters proxied from original method" do
+        is_expected.to eq(expected_parameters)
+      end
+    end
+
+    context "when #unmemoized_method" do
+      let(:method_name) { :unmemoized_method }
+      let(:expected_parameters) { [] }
+
+      it "returns parameters proxied from original method" do
```
Comment (ms-ati) [inline spec/proxying_original_method_params_spec.rb:42]:
```suggestion
      it "returns parameters directly from the method" do
```

**Comment by ms-ati** [comment]
Re: #165

File: lib/memo_wise.rb [hunk 2]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
           target.singleton_class.send(:private, method_name)
         end
       end
+
+      # Override [Module#instance_method](https://ruby-doc.org/core-3.0.0/Module.html#method-i-instance_method)
+      # to proxy the original `UnboundMethod#parameters` results. This proxying
+      # helps code that uses Ruby reflection over method parameters
+      def target.instance_method(symbol)
```
Comment (JacobEvelyn) [inline lib/memo_wise.rb:430]: Whoa, I wasn't aware you could use `def` like this.

File: lib/memo_wise.rb [hunk 3]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
...
+        original_memo_wised_name = :"_memo_wise_original_#{symbol}"
+        curr_method = super
+
+        orig_method =
+          if symbol == :initialize
+            curr_method.super_method
```
Comment (JacobEvelyn) [inline lib/memo_wise.rb:436]: The fact that this evaluates to `super.super_method` is kind of confusing to me. Would you mind adding some comments here for what this is doing and why we need to special-case it?

File: lib/memo_wise.rb [hunk 4]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
...
+          elsif private_method_defined?(original_memo_wised_name)
+            super(original_memo_wised_name)
+          end
+
+        unless orig_method.nil?
```
Comment (JacobEvelyn) [inline lib/memo_wise.rb:441]: Could you add a comment explaining when `orig_method` will be `nil`? Is it just when there's no `initialize` method defined?

File: lib/memo_wise.rb [hunk 5]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
...
+          orig_params = orig_method.parameters
+          curr_method.define_singleton_method(:parameters) { orig_params }
```
Comment (JacobEvelyn) [inline lib/memo_wise.rb:443]: This approach makes sense to me, but I just want to make sure we're on the same page that we think this is better than just proxying the whole `UnboundMethod` object instead.

**Review by JacobEvelyn — CHANGES_REQUESTED** [review]

File: lib/memo_wise.rb (hunk 3 above)
Comment (ms-ati) [reply lib/memo_wise.rb:436]: I'll add explicit documentation of the three paths:

1.  `initialize` - call `super` which is the original implementation of `Module#instance_method`, get back the `UnboundMethod` which represents our `MemoWise` version of `initialize` -- then ask **it** for `#super_method`, because it know the original method that it is overriding
 
2. Any memo-wised method - we need to look up the original method by it's renamed symbol, and call `super` which is the original implementation of `instance_method`, using that name

3. Anything else - we have `orig_method = nil`

File: lib/memo_wise.rb (hunk 4 above)
Comment (ms-ati) [reply lib/memo_wise.rb:441]: I'll clarify the cases -- just returning `curr_method` when it was not a memo_wised method and not initialize -- can make it clearer

File: lib/memo_wise.rb (hunk 5 above)
Comment (ms-ati) [reply lib/memo_wise.rb:443]: Discussed live - I'm open on this - was leaning towards minimal "lying" to start to avoid unintentionally breaking things

**Review by JacobEvelyn — APPROVED** [review]
LGTM if you add more documentation.

File: spec/proxying_original_method_params_spec.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      end
+
+      it "proxies UnboundMethod#parameters via singleton method" do
+        expect(unbound_method.singleton_methods).to eq [:parameters]
+      end
+    end
+
+    context "when #unmemoized_method" do
+      let(:method_name) { :unmemoized_method }
+      let(:expected_parameters) { [] }
```
Comment (jemmaissroff) [inline spec/proxying_original_method_params_spec.rb:50]: Should we have a test case with `unmemoized_methods` and parameters?

File: lib/memo_wise.rb [hunk 7]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
           target.singleton_class.send(:private, method_name)
         end
       end
+
+      # Override [Module#instance_method](https://ruby-doc.org/core-3.0.0/Module.html#method-i-instance_method)
+      # to proxy the original `UnboundMethod#parameters` results. This proxying
+      # helps code that uses Ruby reflection over method parameters
+      def target.instance_method(symbol)
+        original_memo_wised_name = :"_memo_wise_original_#{symbol}"
```
Comment (jemmaissroff) [inline lib/memo_wise.rb:431]: Should we pull this out since we use it elsewhere too? Wouldn't want the case where we change it somewhere and forget to change it here

File: lib/memo_wise.rb [hunk 8]
```diff
@@ -423,6 +423,28 @@ def #{method_name}#{args_str}
...
+        curr_method = super
+
```
Comment (jemmaissroff) [inline lib/memo_wise.rb:433]: I found the below logic a little hard to follow. Wonder what you think of a refactor something like this:

```
        super.tap do |curr_method|
          if symbol == :initialize
            curr_method.define_singleton_method(:parameters) do
              curr_method.super_method.parameters
            end
          elsif private_method_defined?(original_memo_wised_name)
            orig_super = super(original_memo_wised_name)
            curr_method.define_singleton_method(:parameters) do
              orig_super.parameters
            end
          end
        end
```

I think it's cleaner because surrounding the whole thing in a tap makes it explicit that the goal of this method is to amend `super`. It's also more clear to me this way that if the method doesn't fall into either of the cases, we don't amend the method at all.

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: lib/memo_wise.rb (hunk 8 above)
Comment (ms-ati) [reply lib/memo_wise.rb:433]: Sure, I'm happy to refactor this way!

File: lib/memo_wise.rb (hunk 7 above)
Comment (ms-ati) [reply lib/memo_wise.rb:431]: Can we delay that until a following PR when I extract all the private API stuff? I don't want to add additional helper methods now, as they pollute the method namespace of the target classes in the current implementation - I'm planning to change that soon -- and had already responded to you on different PRs that we could start extracting more, smaller helper methods at that point?

Comment (jemmaissroff) [reply lib/memo_wise.rb:431]: Sounds good - sorry I'm losing track a bit of which PRs are still coming in this series and when we're extracting the helpers. As long as you have a good grasp of when we're doing this, works for me!

Comment (ms-ati) [reply lib/memo_wise.rb:431]: Added a TODO for future extraction of the common method naming pattern

File: lib/memo_wise.rb (hunk 3 above)
Comment (ms-ati) [reply lib/memo_wise.rb:436]: Modified per @jemmaissroff above and documented per this discussion, please re-review?

File: lib/memo_wise.rb (hunk 8 above)
Comment (ms-ati) [reply lib/memo_wise.rb:433]: Modifed as discussed, and added comments per @JacobEvelyn below, please re-review?

File: lib/memo_wise.rb (hunk 4 above)
Comment (ms-ati) [reply lib/memo_wise.rb:441]: Please re-review

File: lib/memo_wise.rb (hunk 5 above)
Comment (ms-ati) [reply lib/memo_wise.rb:443]: Agreed to go with just proxying the original parameter values, for minimally-intrusive change

File: spec/proxying_original_method_params_spec.rb (hunk 6 above)
Comment (ms-ati) [reply spec/proxying_original_method_params_spec.rb:50]: Added! Please re-review

File: lib/memo_wise.rb [hunk 9]
```diff
@@ -404,7 +404,7 @@ def #{method_name}#{args_str}
     end
 
     unless target.singleton_class?
-      # Create certain class methods to implement .preset_memo_wise
+      # Create class methods to implement .preset_memo_wise and .reset_memo_wise
```
Comment (ms-ati) [inline lib/memo_wise.rb:407]: Unrelated comment clarification

File: lib/memo_wise.rb [hunk 10]
```diff
@@ -425,25 +425,38 @@ def #{method_name}#{args_str}
       end
 
       # Override [Module#instance_method](https://ruby-doc.org/core-3.0.0/Module.html#method-i-instance_method)
-      # to proxy the original `UnboundMethod#parameters` results. This proxying
-      # helps code that uses Ruby reflection over method parameters
+      # to proxy the original `UnboundMethod#parameters` results. We want the
+      # parameters to reflect the original method in order to support callers
+      # who want to use Ruby reflection to process the method parameters,
+      # because the generated memoized method would have a generic set of
```
Comment (JacobEvelyn) [inline lib/memo_wise.rb:431]: Would you mind tweaking this to say something like `... because our overridden `initialize` method, in some cases the generated memoized methods, would have ...`? (Not sure if you're aware but avoiding generic params like `*args` in many cases is how @jemmaissroff and I got some big speedups recently.)

**Review by jemmaissroff — APPROVED** [review]

File: lib/memo_wise.rb (hunk 8 above)
Comment (jemmaissroff) [reply lib/memo_wise.rb:437]: Looks great, thanks!

File: lib/memo_wise.rb (hunk 10 above)
Comment (ms-ati) [reply lib/memo_wise.rb]: Changed as requested!
